ecosystem/indexer/src/indexer/transaction_processor.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{
+        GOT_CONNECTION, PROCESSOR_ERRORS, PROCESSOR_INVOCATIONS, PROCESSOR_OKS,
+        UNABLE_TO_GET_CONNECTION,
+    },
+    database::{execute_with_better_error, PgDbPool, PgPoolConnection},
+    indexer::{errors::TransactionProcessingError, processing_result::ProcessingResult},
+    models::processor_statuses::ProcessorStatusModel,
+    schema,
+};
+use aptos_rest_client::Transaction;
+use async_trait::async_trait;
+use diesel::{prelude::*, RunQueryDsl};
+use schema::processor_statuses::{self, dsl};
+use std::{fmt::Debug, sync::Arc};
+
+#[async_trait]
+pub trait TransactionProcessor: Send + Sync + Debug {
+    /// name of the processor, for status logging
+    fn name(&self) -> &'static str;
+
+    /// Accepts a transaction, and processes it. This method will be called from `process_transaction_with_status`
+    /// In case a transaction cannot be processed, returns an error.
+    async fn process_transaction(
+        &self,
+        transaction: Arc<Transaction>,
+    ) -> Result<ProcessingResult, TransactionProcessingError>;
+
+    /// Gets a reference to the connection pool
+    fn connection_pool(&self) -> &PgDbPool;
+
+    /// Gets the connection.
+    /// If it was unable to do so (default timeout: 30s), it will keep retrying until it can.
+    fn get_conn(&self) -> PgPoolConnection {
```

Comment:
(Note: this comment stems from my dislike of making trait interfaces super large. I always prefer to separate out what's required in the trait vs. common utility functions that be applied to a specific trait implementation).

nit: from a readability standpoint, everything below here could be a utility helper function not defined within the trait itself, e.g., `transaction_processor_utils.rs`. For example, I could call all of the functions below with a reference to a generic trait implementation of `TransactionProcessor` (and also some of them with a connection instance only). 

My feeling is that these are clearer as utility functions, unless you see very specific examples of where you'd like them to be overridden by other trait implementations. I'm not sure that's the case here.

Reply:
I agree, they could all very well be helper methods instead of instance methods on the trait. I personally like that everything I can do with the trait is already on the trait, instead of having to find other methods somewhere, and it feels much more solidly OOP (to me). Unless you feel very strongly about it I'd leave it as is for now

Reply:
oops + rust lulz